zebra-state/src/in_memory/chains.rs

```diff
@@ -3,18 +3,98 @@ use crate::{Request, Response};
 use futures::prelude::*;
 use std::{
     pin::Pin,
+    sync::Arc,
     task::{Context, Poll},
 };
-use tower::Service;
+use tower::{Service, ServiceExt};
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+/// A representation of the chain state at a given height
+#[derive(Clone, Debug)]
+struct ChainState {
+    block: Arc<Block>,
```

Comment:
Do we actually need the block here?

If it's already stored in the inner state, then we're storing it twice, which could lead to desynchronisation bugs.

Reply:
We could just store the block hash.

In general, we should split the fields we need out into im data structures, rather than keeping a reference to the entire block. It will be faster to access, and easier to code.

Reply:
I wasn't planning on storing it in the state

Reply:
I'm not sure what the best design is here.

If we store the blocks outside the chain context, then we could accidentally drop a block, but keep its verification context.

If we store the blocks in the chain context, then the data we use for verification is less explicit. There could also be a slightly higher cost to access recent blocks, rather than a specific data structure. But I don't think it will matter in practice.

I think storing data in multiple places is actually a greater risk.